frappe/types/filter.py

```diff
@@ -0,0 +1,280 @@
+		return "UNSPECIFIED"
+
+
+UNSPECIFIED = Sentinel()
+
+T = TypeVar("T")
+
+
+def is_not_unspecified(value: T | Sentinel) -> TypeGuard[T]:
+	return value is not UNSPECIFIED
+
+
+class _FilterTuple(NamedTuple):
+	doctype: Doct
+	fieldname: Fld
+	operator: Op
+	value: Val
+
+
+def _type_narrow(v: _InVal) -> _Val:
+	if isinstance(v, bool):  # beware: bool derives int in _Val
+		return int(v)
+	elif isinstance(v, _Val):
+		return v
+	elif isinstance(v, DocRef):  # type: ignore[redundant-expr]
+		return v.__value__()
+	else:
+		raise ValueError(
+			f"Value must be one of types: {', '.join(str(t.__name__) for t in _InVal.__args__)}; found {type(v)}"
+		)
+
+
+class FilterTuple(_FilterTuple):
+	"""A named tuple representing a filter condition."""
+
+	def __new__(
+		cls,
+		s: FilterTupleSpec | None = None,
+		/,
+		*,
+		doctype: Doct | Sentinel = UNSPECIFIED,
+		fieldname: Fld | Sentinel = UNSPECIFIED,
+		operator: Op = "=",
+		value: InVal | Sentinel = UNSPECIFIED,
+	) -> Self:
+		"""
+		Create a new FilterTuple instance.
+		Args:
+		        s: A sequence representing the filter tuple.
+		        doctype: The document type.
+		        fieldname: The field name.
+		        operator: The comparison operator.
+		        value: The value to compare against.
+		Returns:
+		        A new FilterTuple instance.
+		"""
+		try:
+			if isinstance(s, Sequence):
+				if len(s) == 2:
+					fieldname, value = s
+				elif len(s) == 3:
+					fieldname, operator, value = s
+				elif len(s) == 4:  # type: ignore[redundant-expr]
+					doctype, fieldname, operator, value = s
+				else:
+					raise ValueError(f"Invalid sequence length: {len(s)}. Expected 2, 3, or 4 elements.")
+			if not is_not_unspecified(doctype) or doctype is None:
+				raise ValueError("doctype is required")
+			if not is_not_unspecified(fieldname) or fieldname is None:
+				raise ValueError("fieldname is required")
+			if not is_not_unspecified(value):
+				raise ValueError("value is required; can be None")
+
+			# soundness
+			if operator in ("in", "not in") and isinstance(value, str):
+				value = value.split(",")
+
+			_value: Val
+			if isinstance(value, _InVal):
+				_value = _type_narrow(value)
+			else:
+				_value = tuple(_type_narrow(v) for v in value)
+
+			return super().__new__(
+				cls,
+				doctype=doctype,
+				fieldname=fieldname,
+				operator=operator,
+				value=_value,
+			)
+
+		except Exception as e:
+			error_context = (
+				f"Error creating FilterTuple:\n"
+				f"Input: {s}, doctype={doctype}, fieldname={fieldname}, operator={operator}, value={value}\n"
+				f"Error: {e!s}\n"
+				f"Usage: FilterTuple( (fieldname, value),                  doctype=dt )\n"
+				f"       FilterTuple( (fieldname, operator, value),        doctype=dt )\n"
+				f"       FilterTuple( (doctype, fieldname, operator, value)           )\n"
+				f"       FilterTuple( doctype=doctype, fieldname=fieldname, operator=operator, value=value )"
+			)
+			raise ValueError(error_context) from e
```

Comment:
This is not a user facing error and consistent with prior behavior: exceptions due to malformed filters haven't been cought across implementation sites.